man/man8/puppetd.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPET\-AGENT" "8" "June 2012" "Puppet Labs, LLC" "Puppet manual"
.
.SH "NAME"
\fBpuppet\-agent\fR \- The puppet agent daemon
.
.SH "SYNOPSIS"
Retrieves the client configuration from the puppet master and applies it to the local host\.
.
.P
This service may be run as a daemon, run periodically using cron (or something similar), or run interactively for testing purposes\.
.
.SH "USAGE"
puppet agent [\-\-certname \fIname\fR] [\-D|\-\-daemonize|\-\-no\-daemonize] [\-d|\-\-debug] [\-\-detailed\-exitcodes] [\-\-digest \fIdigest\fR] [\-\-disable] [\-\-enable] [\-\-fingerprint] [\-h|\-\-help] [\-l|\-\-logdest syslog|\fIfile\fR|console] [\-\-no\-client] [\-\-noop] [\-o|\-\-onetime] [\-\-serve \fIhandler\fR] [\-t|\-\-test] [\-v|\-\-verbose] [\-V|\-\-version] [\-w|\-\-waitforcert \fIseconds\fR]
.
.SH "DESCRIPTION"
This is the main puppet client\. Its job is to retrieve the local machine\'s configuration from a remote server and apply it\. In order to successfully communicate with the remote server, the client must have a certificate signed by a certificate authority that the server trusts; the recommended method for this, at the moment, is to run a certificate authority as part of the puppet server (which is the default)\. The client will connect and request a signed certificate, and will continue connecting until it receives one\.
.
.P
Once the client has a signed certificate, it will retrieve its configuration and apply it\.
.
.SH "USAGE NOTES"
\'puppet agent\' does its best to find a compromise between interactive use and daemon use\. Run with no arguments and no configuration, it will go into the background, attempt to get a signed certificate, and retrieve and apply its configuration every 30 minutes\.
.
.P
Some flags are meant specifically for interactive use \-\- in particular, \'test\', \'tags\' or \'fingerprint\' are useful\. \'test\' enables verbose logging, causes the daemon to stay in the foreground, exits if the server\'s configuration is invalid (this happens if, for instance, you\'ve left a syntax error on the server), and exits after running the configuration once (rather than hanging around as a long\-running process)\.
.
.P
\'tags\' allows you to specify what portions of a configuration you want to apply\. Puppet elements are tagged with all of the class or definition names that contain them, and you can use the \'tags\' flag to specify one of these names, causing only configuration elements contained within that class or definition to be applied\. This is very useful when you are testing new configurations \-\- for instance, if you are just starting to manage \'ntpd\', you would put all of the new elements into an \'ntpd\' class, and call puppet with \'\-\-tags ntpd\', which would only apply that small portion of the configuration during your testing, rather than applying the whole thing\.
.
.P
\'fingerprint\' is a one\-time flag\. In this mode \'puppet agent\' will run once and display on the console (and in the log) the current certificate (or certificate request) fingerprint\. Providing the \'\-\-digest\' option allows to use a different digest algorithm to generate the fingerprint\. The main use is to verify that before signing a certificate request on the master, the certificate request the master received is the same as the one the client sent (to prevent against man\-in\-the\-middle attacks when signing certificates)\.
.
.SH "OPTIONS"
Note that any configuration parameter that\'s valid in the configuration file is also a valid long argument\. For example, \'server\' is a valid configuration parameter, so you can specify \'\-\-server \fIservername\fR\' as an argument\.
.
.P
See the configuration file documentation at http://docs\.puppetlabs\.com/references/stable/configuration\.html for the full list of acceptable parameters\. A commented list of all configuration options can also be generated by running puppet agent with \'\-\-genconfig\'\.
.
.TP
\-\-certname
Set the certname (unique ID) of the client\. The master reads this unique identifying string, which is usually set to the node\'s fully\-qualified domain name, to determine which configurations the node will receive\. Use this option to debug setup problems or implement unusual node identification schemes\.
.
.TP
\-\-daemonize
Send the process into the background\. This is the default\.
.
.TP
\-\-no\-daemonize
Do not send the process into the background\.
.
.TP
\-\-debug
Enable full debugging\.
.
.TP
\-\-detailed\-exitcodes
Provide transaction information via exit codes\. If this is enabled, an exit code of \'2\' means there were changes, an exit code of \'4\' means there were failures during the transaction, and an exit code of \'6\' means there were both changes and failures\.
.
.TP
\-\-digest
Change the certificate fingerprinting digest algorithm\. The default is MD5\. Valid values depends on the version of OpenSSL installed, but should always at least contain MD5, MD2, SHA1 and SHA256\.
.
.TP
\-\-disable
Disable working on the local system\. This puts a lock file in place, causing \'puppet agent\' not to work on the system until the lock file is removed\. This is useful if you are testing a configuration and do not want the central configuration to override the local state until everything is tested and committed\.
.
.IP
\'puppet agent\' uses the same lock file while it is running, so no more than one \'puppet agent\' process is working at a time\.
.
.IP
\'puppet agent\' exits after executing this\.
.
.TP
\-\-enable
Enable working on the local system\. This removes any lock file, causing \'puppet agent\' to start managing the local system again (although it will continue to use its normal scheduling, so it might not start for another half hour)\.
.
.IP
\'puppet agent\' exits after executing this\.
.
.TP
\-\-fingerprint
Display the current certificate or certificate signing request fingerprint and then exit\. Use the \'\-\-digest\' option to change the digest algorithm used\.
.
.TP
\-\-help
Print this help message
.
.TP
\-\-logdest
Where to send messages\. Choose between syslog, the console, and a log file\. Defaults to sending messages to syslog, or the console if debugging or verbosity is enabled\.
.
.TP
\-\-no\-client
Do not create a config client\. This will cause the daemon to run without ever checking for its configuration automatically, and only makes sense when puppet agent is being run with listen = true in puppet\.conf or was started with the \fB\-\-listen\fR option\.
.
.TP
\-\-noop
Use \'noop\' mode where the daemon runs in a no\-op or dry\-run mode\. This is useful for seeing what changes Puppet will make without actually executing the changes\.
.
.TP
\-\-onetime
Run the configuration once\. Runs a single (normally daemonized) Puppet run\. Useful for interactively running puppet agent when used in conjunction with the \-\-no\-daemonize option\.
.
.TP
\-\-serve
Start another type of server\. By default, \'puppet agent\' will start a service handler that allows authenticated and authorized remote nodes to trigger the configuration to be pulled down and applied\. You can specify any handler here that does not require configuration, e\.g\., filebucket, ca, or resource\. The handlers are in \'lib/puppet/network/handler\', and the names must match exactly, both in the call to \'serve\' and in \'namespaceauth\.conf\'\.
.
.TP
\-\-test
Enable the most common options used for testing\. These are \'onetime\', \'verbose\', \'ignorecache\', \'no\-daemonize\', \'no\-usecacheonfailure\', \'detailed\-exit\-codes\', \'no\-splay\', and \'show_diff\'\.
.
.TP
\-\-verbose
Turn on verbose reporting\.
.
.TP
\-\-version
Print the puppet version number and exit\.
.
.TP
\-\-waitforcert
This option only matters for daemons that do not yet have certificates and it is enabled by default, with a value of 120 (seconds)\. This causes \'puppet agent\' to connect to the server every 2 minutes and ask it to sign a certificate request\. This is useful for the initial setup of a puppet client\. You can turn off waiting for certificates by specifying a time of 0\.
.
.SH "EXAMPLE"
.
.nf

$ puppet agent \-\-server puppet\.domain\.com
.
.fi
.
.SH "DIAGNOSTICS"
Puppet agent accepts the following signals:
.
.TP
SIGHUP
Restart the puppet agent daemon\.
.
.TP
SIGINT and SIGTERM
Shut down the puppet agent daemon\.
.
.TP
SIGUSR1
Immediately retrieve and apply configurations from the puppet master\.
.
.SH "AUTHOR"
Luke Kanies
.
.SH "COPYRIGHT"
Copyright (c) 2011 Puppet Labs, LLC Licensed under the Apache 2\.0 License


CHANGELOG
3.0.0rc3
===
e619304 (#14761) Add boot, reboot to excludes list for redhat provider
7c1d8c2 (maint) Disable upstart spec test on windows
37df5c4 (#14615) Final stub fixes for the tests.
f11dfd6 (#14615) Fix provider spec tests to work on non gentoo systems
060810e maint: Make it easier to test the service provider
c0cb0fa (Maint) Close filehandles after read in upstart
55505db (Maint) Fix upstart provider to work on Ruby 1.9.3
b0f1d58 Update redhat.spec dependency to include hiera-puppet
6e2f23d Update puppet gem dependencies
d19f93f Update redhat init scripts to use `puppet (agent|queue|master)`
a45942b (#14297) Don't remove tty1 or rsyslog during upstart acceptance
466ab49 (#14297) Define agent in upstart acceptance test
f5d1f74 (#14297) Update upstart provider acceptance tests
3c5ef40 (#14297) Clean up upstart provider and tests
36c41e1 (#14343) Test cases for more forms of variable lookup
6cca061 (#14297) Add acceptance test for upstart provider
b9ed026 (#14297) Update spec tests for upstart provider
91827e8 (#14297) Handle upstart services better
3541db6 (#14343) Lookup in correct inherited scope
2d83e22 (Maint) Added test for fully qualified variables
0788ea2 (#14615) Exclude helperscripts in gentoo service provider
2864d86 (maint) Change Facter dep from exit to raise
2ec9b74 (maint) Make Puppet have hard dependency on Facter 2.0
6c37538 (#3909) Move default certname logic to Puppet::Settings
6be09f0 (#3909) Strip trailing dot from default FQDN

3.0.0rc2
===
0cea47e (#14609) Fix master initialization when running under Passenger
b6c5e9d (#14514) Check for the presence of hiera config
ca5271f (maint) Add data binding Hiera tests

3.0.0rc1
===
a731e15 Update CHANGELOG lib/puppet.rb puppet.spec for 2.7.15rc1
53e0057 Use internal domain name for module tool tests
f4ba59e Use absolute_path in module install spec
ead9d25 (Maint) Color console output when the win32console gem is installed
a76d873 (#14424) Expand path of the target directory
95b75e1 Fix unit test failure in fileserver due to modules
da5361d Maint: De-genericize Puppet Forge references in module face
3b144bf Revert "maint: use relative paths to fixtures."
dae5539 (#14440) Hack to allow help faces to be used as API
e862cad Fix broken man page rake task
b26699a (#10146) `-` is not legal in variable names.
7ce9e17 maint: use relative paths to fixtures.
a7a532c Significantly improve compilation performance when using modules
12c5bda Fix a stub of ansicolor that I missed
ea740f6 Add Puppet::Util::Platform to abstract platform checks
91a4a7d Don't use features to check for color in logging
80fe59d Remove a bad test that can lead to random hangs...
0efb4c2 (#14296) Improve template function error message.
d5e0cdd (#14288) gem provider should tolerate bad lines.
574e4c8 Update CHANGELOG, spec files for 2.7.14
d2b3ba0 Revert "Fix incorrect argument handling for expire in NodeExpirer"
7c7d4b6 Fix #14123 for Windows
1c9bec0 Fix incorrect argument handling for expire in NodeExpirer
7e92a7c Fix filebucket specs on Windows
a713f07 (#8778) Make '' == undef commutative in the DSL
24d4bb5 Revert "(#11004) Fix regex for pkg provider line matching"
34da6f7 (#13651) Use magenta, not purple in Facter::Util::Colors
e37b536 (#11004) Fix regex for pkg provider line matching
9b47110 (#13956) Fix failing spec tests
7e92dcb Just change Puppet settings in tests, rather than stubbing.
ca313dc (maint) Fix destinations_spec failure on windows
cde3bee Just change Puppet settings in tests, rather than stubbing.
6aa8441 (#11004) Support Solaris 11 output in pkg package provider
61e7d3c maint: Fix pkg_spec test
603b368 (#14173) Enforce that filebucket paths must be absolute
b7828ed (#3452) Autorequire user for cron
a47c42d Update CHANGELOG for 2.7.14rc3
402a425 (#14127) Add integration tests for ssh_authorized_key
85ee441 maint: refactor integration specs for ssh_authorized_key
42b20be (#13682) Fix acceptance test failures
dce3c32 (#13682) Rename Puppet::Module::Tool to Puppet::ModuleTool
5117e12 (maint) Ensure every file has a trailing newlines
b4d1c65 (#14127) ssh_authorized_keys grammer fails on blank lines.
25ad6d7 Updating CHANGELOG for Puppet 2.7.14rc2
74e462f (#14060) Fix quoting of commands to interpolate inside the shell.
796a072 Modernize the style of the shell exec provider tests.
9e526c9 (#14060) Fix quoting of commands to interpolate inside the shell.
36d959a (#14123) Puppet shouldn't explode if PATH contains ~nonexistent_user
a009c54 Modernize the style of the shell exec provider tests.
934ec89 Improve deprecation warning for dynamic lookup
c255414 (#14036) Add upstart tests
dc3bb9f (#14036) Handle upstart better
a281c48 (#13645) Add test cases to tagmail_spec
3a4ac60 Updating CHANGELOG, lib/puppet.rb for 2.7.14rc1
f6ac1bb (#13645) Do not open a smtp connection if nothing to report
f40cc71 (#12392): Colorize console output on Windows
0d9e852 Removed duplication of compiling a catalog
d67711d Cleaner test for create_resources doing nothing
2d3cc63 Fixing problem caused by activerecord
f1331ba Put back require 'puppet' in Puppet::Util
79bf4fa (#13299) Fix error message when agent is disabled
5178adc Update puppet.spec for rpm packaging
fb96747 Fixed old log test to match new autoflush behavior
878127f Make sure the log file writable
72d3b56 Default autoflushing of log files to true
9711d50 Created test that shows enc warning problem
d54100d (Maint) Fix order-dependent test failure
f531400 (Maint) Don't assume eventlog gem is installed on Windows
2957ffe Fix tempfile cleanup in specs
e2d6915 Fix communication_error acceptance test
27a5eb7 options[:ca_location] is a string not a symbol. Safer to pull Puppet::SSL::Host.ca_location since it is already set correctly.
c01cac9 (#13204) Don't ignore missing PATH.augnew files
7d27cd2 Use :as_platform in the smf service spec
876785a Use :as_platform in the redhat service spec

2.7.16rc1
===
3a06246 (#8858) Explicitly set SSL peer verification mode.
cd4bee8 Test SSL setup code, not our stubbing of it.

lib/puppet/application/puppetd.rb
require 'puppet'
require 'puppet/application'
require 'puppet/agent'
require 'puppet/daemon'
require 'puppet/configurer'
require 'puppet/network/client'

Puppet::Application.new(:puppetd) do

    should_parse_config

    attr_accessor :explicit_waitforcert, :args, :agent, :daemon

    preinit do
        # Do an initial trap, so that cancels don't get a stack trace.
        trap(:INT) do
            $stderr.puts "Cancelling startup"
            exit(0)
        end

        {
            :waitforcert => 120,  # Default to checking for certs every 5 minutes
            :onetime => false,
            :verbose => false,
            :debug => false,
            :centrallogs => false,
            :setdest => false,
            :enable => false,
            :disable => false,
            :client => true,
            :fqdn => nil,
            :serve => []
        }.each do |opt,val|
            options[opt] = val
        end

        @explicit_waitforcert = false
        @args = {}
        @daemon = Puppet::Daemon.new
        @daemon.argv = ARGV.dup
    end

    option("--centrallogging")
    option("--disable")
    option("--enable")
    option("--debug","-d")
    option("--fqdn FQDN","-f")
    option("--test","-t")
    option("--verbose","-v")

    option("--serve HANDLER", "-s") do |arg|
        if Puppet::Network::Handler.handler(arg)
            options[:serve] << arg.to_sym
        else
            raise "Could not find handler for %s" % arg
        end
    end

    option("--no-client") do |arg|
        options[:client] = false
    end

    option("--onetime", "-o") do |arg|
        options[:onetime] = true
        options[:waitforcert] = 0 unless @explicit_waitforcert
    end

    option("--logdest DEST", "-l DEST") do |arg|
        begin
            Puppet::Util::Log.newdestination(arg)
            options[:setdest] = true
        rescue => detail
            if Puppet[:debug]
                puts detail.backtrace
            end
            $stderr.puts detail.to_s
        end
    end

    option("--waitforcert WAITFORCERT", "-w") do |arg|
        options[:waitforcert] = arg.to_i
        @explicit_waitforcert = true
    end

    option("--port PORT","-p") do |arg|
        @args[:Port] = arg
    end

    dispatch do
        return :onetime if options[:onetime]
        return :main
    end

    command(:onetime) do
        unless options[:client]
            $stderr.puts "onetime is specified but there is no client"
            exit(43)
        end

        @daemon.set_signal_traps

        begin
            @agent.run
        rescue => detail
            if Puppet[:trace]
                puts detail.backtrace
            end
            Puppet.err detail.to_s
        end
        exit(0)
    end

    command(:main) do
        Puppet.notice "Starting Puppet client version %s" % [Puppet.version]

        @daemon.start
    end

    # Enable all of the most common test options.
    def setup_test
        Puppet.settings.handlearg("--ignorecache")
        Puppet.settings.handlearg("--no-usecacheonfailure")
        Puppet.settings.handlearg("--no-splay")
        Puppet.settings.handlearg("--show_diff")
        Puppet.settings.handlearg("--no-daemonize")
        options[:verbose] = true
        options[:onetime] = true
        options[:waitforcert] = 0 unless @explicit_waitforcert
    end

    # Handle the logging settings.
    def setup_logs
        if options[:debug] or options[:verbose]
            Puppet::Util::Log.newdestination(:console)
            if options[:debug]
                Puppet::Util::Log.level = :debug
            else
                Puppet::Util::Log.level = :info
            end
        end

        unless options[:setdest]
            Puppet::Util::Log.newdestination(:syslog)
        end
    end

    def enable_disable_client(agent)
        if options[:enable]
            agent.enable
        elsif options[:disable]
            agent.disable
        end
        exit(0)
    end

    def setup_listen
        unless FileTest.exists?(Puppet[:authconfig])
            Puppet.err "Will not start without authorization file %s" %
                Puppet[:authconfig]
            exit(14)
        end

        handlers = nil

        if options[:serve].empty?
            handlers = [:Runner]
        else
            handlers = options[:serve]
        end

        require 'puppet/network/server'
        # No REST handlers yet.
        server = Puppet::Network::Server.new(:xmlrpc_handlers => handlers, :port => Puppet[:puppetport])

        @daemon.server = server
    end

    setup do
        setup_test if options[:test]

        setup_logs

        if Puppet.settings.print_configs?
            exit(Puppet.settings.print_configs ? 0 : 1)
        end

        # If noop is set, then also enable diffs
        if Puppet[:noop]
            Puppet[:show_diff] = true
        end

        args[:Server] = Puppet[:server]
        if options[:fqdn]
            args[:FQDN] = options[:fqdn]
            Puppet[:certname] = options[:fqdn]
        end

        if options[:centrallogs]
            logdest = args[:Server]

            if args.include?(:Port)
                logdest += ":" + args[:Port]
            end
            Puppet::Util::Log.newdestination(logdest)
        end

        Puppet.settings.use :main, :puppetd, :ssl

        # This configures all of the SSL-related indirected classes.
        Puppet::SSL::Host.ca_location = :remote

        Puppet::Transaction::Report.terminus_class = :rest

        # Override the default; puppetd needs this, usually.
        # You can still override this on the command-line with, e.g., :compiler.
        Puppet[:catalog_terminus] = :rest

        # Override the default.
        Puppet[:facts_terminus] = :facter

        Puppet::Resource::Catalog.cache_class = :yaml


        # We need tomake the client either way, we just don't start it
        # if --no-client is set.
        @agent = Puppet::Agent.new(Puppet::Configurer)

        enable_disable_client(@agent) if options[:enable] or options[:disable]

        @daemon.agent = agent if options[:client]

        # It'd be nice to daemonize later, but we have to daemonize before the
        # waitforcert happens.
        if Puppet[:daemonize]
            @daemon.daemonize
        end

        host = Puppet::SSL::Host.new
        cert = host.wait_for_cert(options[:waitforcert])

        @objects = []

        # This has to go after the certs are dealt with.
        if Puppet[:listen]
            unless options[:onetime]
                setup_listen
            else
                Puppet.notice "Ignoring --listen on onetime run"
            end
        end
    end
end
